test/test_tree.py

```diff
@@ -40,6 +40,41 @@ def test_serializable(self):
             testtree._deserialize(stream)
         # END for each item in tree
 
+    def test_tree_modifier_ordering(self):
+        hexsha = "6c1faef799095f3990e9970bc2cb10aa0221cf9c"
+        roottree = self.rorepo.tree(hexsha)
+        blob_mode = Tree.blob_id << 12
+        tree_mode = Tree.tree_id << 12
+
+        files_in_desired_order = [
+            (blob_mode, "_.htaccess"),
+            (tree_mode, "fileadmin"),
+            (blob_mode, "index.php"),
+            (tree_mode, "typo3"),
+            (tree_mode, "typo3_src-6.2.12"),
+            (tree_mode, "typo3_src"),
+            (blob_mode, "typo3cms"),
+            (tree_mode, "typo3conf"),
+            (tree_mode, "uploads"),
+        ]
+        mod = roottree.cache
+        for file_mode, file_name in files_in_desired_order:
+            mod.add(hexsha, file_mode, file_name)
+        # end for each file
+
+        def file_names_in_order():
+            return [t[1] for t in files_in_desired_order]
+
+        def names_in_mod_cache():
+            a = [t[2] for t in mod._cache]
+            here = file_names_in_order()
+            return [e for e in a if e in here]
+
+        assert names_in_mod_cache() == file_names_in_order(), "add() keeps order"
```

Comment:
Could you make this `self.assertEqual()`? That way I'd expect better error message in case of failure, a failure which should show up in CI at least once with the unfixed version of the code.

Reply:
The `pytest` runner [rewrites the bytecode of `assert` statements](https://docs.pytest.org/en/7.1.x/how-to/assert.html) appearing in tests, including in `unittest` tests, to provide the kind of [rich introspection](https://docs.pytest.org/en/7.1.x/example/reportingdemo.html) for `assert lhs == rhs` as for `self.assertEqual(lhs, rhs)` (while still evaluating expressions only once). It is possible to direct `pytest` not to do this, but that is rarely done.

This rich introspection includes element-by-element diffs when `lhs` and `rhs` are lists, as in this test. It is also done even when a custom assertion message is specified, as here: the rich introspection is shown after the message. So I would not expect using `self.assertEqual` would give better information here than `assert`; they should give output of the same quality. Whether or not this change is to be made, *if* it makes a difference in how much information is shown on failure, then that seems like a separate bug that should be investigated.

Experiments seem to bear this out, and suggest this change may not be needed: [a failure with the original test using `assert`](https://github.com/EliahKagan/GitPython/actions/runs/7504730404/job/20432447621#step:12:536) provides as detailed list introspection as [a failure when the test uses `self.assertEqual`](https://github.com/EliahKagan/GitPython/actions/runs/7504776269/job/20432592653#step:12:536). (I've linked to the part of each output that shows the first report, since that's where the test code can be examined along with the results. But the rich introspection can also be seen, in both cases, in the summaries at the end of the runs.)

Of course, that is all with the `pytest` runner. `unittest.TestCase` assertion methods should still be used if `unittest` tests are going to be run with the `unittest` runner rather than the `pytest` runner, and a few of the modules are possible to run fully with the `unittest` runner, including the `test_tree.py` module at issue here. Overall, though, GitPython's test suite is not compatible with the `unittest` runner, because some tests are pure pytest tests, but even more so because various `pytest` features are used throughout the test suite, in `unittest` tests and their decorations. (One of the goals of `pytest` is to allow its features to be gradually introduced in tests written against other frameworks that the `pytest` runner supports.)

Currently the `test_tree.py` module makes little use of `unittest.TestCase` assertion methods, already using `assert` statements instead in most cases, including for all equality assertions. So if this assertion is to be changed from the `assert lhs == rhs, message` form to the `self.assertEqual(lhs, rhs, message)` form, then I think it may be best to change the other assertions throughout that test module as well.

Reply:
Thanks for the clarification! Under these circumstances that change request is definitely void, after all, the current assert style is more readable and lightweight.

I think it would be interesting to one day unify the assert-style within the project, as it will then just be about consistency, not about features.